### ISSUE TITLE
(NULL == "") == true ?

### ISSUE THREAD

**Issue by nigoroll** [varnishcache/varnish-cache#3844]
I am pretty sure this has been discussed before, but I stumbled over it again and would like to revisit because I experienced it to be very un-pola:

We print a C `NULL` string as `""`, but a comparison with `""` yields false.

**Comment by gquintard** [comment]
agreed, it's fine for C devs, but it's confusing for the rest of the population. Changing that is a major breaking change though, all the VCLs would need to at least be reviewed and possibly updated

**Comment by Dridi** [comment]
As long as a BOOL evaluation of a HEADER can be used to test existence, I have no problem with `HEADER == STRING` to turn be degraded to `STRING == STRING` (and turning a non existent header into a string would result in an empty string, just like an assignment would).

**Comment by gquintard** [comment]
I think I'd prefer for the HEADER to be cast into a STRING and then into a BOOL, with `""` being false-ish, and have a `std.header_exists(HEADER)` function to actually test presence.

**Comment by nigoroll** [comment]
The case here was with a vmod function returning `NULL`, but I think @gquintard 's suggestion has some merit to it: `NULL` could signify "undefined/unset", but still evaluate to `""` if used as a string, then the following could hold true:
* `std.defined(req.http.inexistent-header) == false`
* `(req.http.inexistent-header == "") == true`

**Comment by nigoroll** [comment]
Summary of bugwash discussion:

The current proposal basically boils down to
* for type conversions/folding, preserve `NULL`: `Typea(NULL) => Typeb(NULL)`
* for string contexts, convert `NULL` to `""`: `String(NULL) => ""`
* for boolean contexts, convert `NULL` to false: `BOOL(NULL) => false`

So far, the rules do not require an exception for comparisons, the [examples](https://github.com/varnishcache/varnish-cache/issues/3844#issuecomment-1243632961) can be implemented with these rules only (and the existing string comparison).

Some aspects relating to `String(NULL) => ""` were discussed:

* If `NULL` handling was to be transitive, `set` would need to have `unset` semantics such that `!req.http.a == !req.http.inexistent-header` holds for
```
set req.http.a = req.http.inexistent-header;
```

* But as long as we want to preserve `String(NULL) => ""`, such that in the example above, header `a` will be set to the empty string (as it is now), we will need to define `BOOL("") == true` such that `BOOL(req.http.empty-header) == true` and `BOOL(req.http.inexistent-header) == false`. This is the case with currecnt code.

It was argued that the empty string should evaluate to `false` such that `BOOL(x) == BOOL(String(x))`.

**Comment by Dridi** [comment]
I tried a light survey of the current situation:

```vcl
varnishtest cmp

varnish v1 -vcl {
	import debug;

	backend be none;

	sub vcl_recv {
		return (synth(200));
	}

	sub vcl_synth {
		set resp.http.noheader-eq-empty = req.http.noheader == "";
		set resp.http.empty-eq-noheader = "" == req.http.noheader;

		set resp.http.nostring-eq-empty = debug.null_string() == "";
		set resp.http.empty-eq-nostring = "" == debug.null_string();

		#set resp.http.nostrands-eq-empty = debug.null_strands() == "";
		#set resp.http.empty-eq-nostrands = "" == debug.null_strands();

		set resp.http.not-empty = !"";

		#set resp.http.nobackend-eq-empty = req.backend_hint == "";
		#set resp.http.empty-eq-nobackend = "" == req.backend_hint;

		#set resp.http.nobackend-eq-false = req.backend_hint == false;
		#set resp.http.false-eq-nobackend = false == req.backend_hint;

		if (req.backend_hint) {
			set resp.http.backend_hint = "true";
		} else {
			set resp.http.backend_hint = "false";
		}
	}
} -start

client c1 {
	txreq
	rxresp

	expect resp.http.noheader-eq-empty == true
	expect resp.http.empty-eq-noheader == true

	expect resp.http.nostring-eq-empty == true
	expect resp.http.empty-eq-nostring == true

	#expect resp.http.nostrands-eq-empty == true
	#expect resp.http.empty-eq-nostrands == true

	expect resp.http.not-empty == true

	#expect resp.http.nobackend-eq-empty == true
	#expect resp.http.empty-eq-nobackend == true

	#expect resp.http.nobackend-eq-false == true
	#expect resp.http.false-eq-nobackend == true

	expect resp.http.backend_hint == false
} -start
```

It relies on non missing vmod_debug trivial functions, and it looks like today we can't even generalize `BOOL(x) == BOOL(String(x))`, because `BOOL(x)` automatic conversion will not happen when prompted by a boolean operator.


Lines commented out currently don't compile without a dummy `"" + x` concatenation, but the commented out results reflect the outcome when that dummy concatenation is present (confirming that none values turn into empty strings, not null strings).

So maybe the first step should be to fill the gap and allow `TYPE cmp_operator STRING|BOOL` to compile, if we consider it a gap.

**Comment by gquintard** [comment]
testing nullness this way work, but my (and Nils' I believe) is that `""` is trueish, would we change that once we can `req.http.foo != none`?
